### PR TITLE
sql: skeletal support for execution outliers

### LINKED ISSUE #79432: Skeletal support for a slow/failed queries view in the SQL CLI
This task captures the initial walking skeleton for query outliers: an in-memory, node-local view, exposed in `crdb_internal.node_execution_outliers`.

At this point, the table will contain only a single row, capturing only the session id, transaction id, and statement id of the most recent statement that took longer than 1 second (or perhaps 100ms? See @kevin-v-ngo for telemetry.) to run. Perhaps gate all of this on a cluster setting, defaulting to disabled. (Can we hide the table completely if so?)

We defer nearly every other concern -- detection, memory management, data collection, persistence, API, UI, etc. -- to future tasks. But this work should include some baseline performance measuring, running the KV95 workload.

Epic: CRDB-13544

Jira issue: CRDB-14853

### PR DESCRIPTION
**Pull request by matthewtodd**
Closes #79432.

This commit introduces a disabled-by-default "outliers" subsystem of
sqlstats, intended to report on statements and transactions that take
longer than usual.

The most interesting (and challenging) parts of this work have been left
for future commits. While we will eventually develop some measure of
statistical significance, at the moment, detection is currently based on
a simple, configurable latency threshold. Only the most recent 10
entries are retained in memory, and there is no persistence and no
inter-node communication.

What is here are the beginnings of a
`crdb_internal.node_execution_outliers` table, listing the session,
transaction, and statement ids of any of these long-running statements,
along with those of any sibling statements within the same transaction.
We will return to add more information to this table and to build a web
UI on top of it.

You may enable outlier detection by setting the
`sql.stats.outliers.experimental.latency_threshold` cluster setting,
much like the slow query log. (As we discover which cluster settings we
want, we will promote them up out of this experimental namespace.)

🚀

Release note: None

### CODE REVIEW

**Comment by matthewtodd** [comment]
Though this outliers subsystem is disabled by default, out of an abundance of caution I've run a kv95 workload on both master and this branch, both with outliers disabled and with outliers enabled with a 1-second latency threshold.

As far as I can tell, there's no risk in landing this branch as it stands, though I would appreciate more eyes on these numbers (and on the workload setup), as I'm not super-familiar with the inner workings here. (Also, is there a better way to get an overview of these stats?)

## Setup
```
cockroach workload init kv --splits 1000 --read-percent 95
cockroach workload run kv --read-percent 95 --concurrency 64 --sequential --duration 30m
```
## On `master`
```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
 1800.0s        0       16652258         9251.3      6.4      5.0     17.8     24.1     96.5  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
 1800.0s        0         877357          487.4      9.6      8.1     23.1     33.6    134.2  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
 1800.0s        0       17529615         9738.7      6.6      5.2     17.8     25.2    134.2  
```

## On `outliers`, with `latency_threshold = 0`
```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
 1800.0s        0       16562946         9201.6      6.4      5.0     17.8     25.2    104.9  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
 1800.0s        0         870715          483.7      9.7      8.1     23.1     32.5    142.6  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
 1800.0s        0       17433661         9685.4      6.6      5.2     17.8     25.2    142.6  
```

## On `outliers`, with `latency_threshold = 1s`
```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
 1800.0s        0       16525525         9180.8      6.5      5.0     17.8     25.2    142.6  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
 1800.0s        0         869210          482.9      9.9      8.4     24.1     33.6    167.8  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
 1800.0s        0       17394735         9663.7      6.6      5.2     17.8     25.2    167.8  
```

**Review by Azhng — COMMENTED** [review]
> Reviewed 15 of 21 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80503)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @matthewtodd)

___
*[`pkg/sql/crdb_internal.go` line 6088 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80503#-N0XIZq12hML7SDfSEQU:-N0XIZq12hML7SDfSEQV:b-5ulfkv) ([raw file](https://github.com/cockroachdb/cockroach/blob/4249409b6f054459c67362c5d6790ac1bcd35421/pkg/sql/crdb_internal.go#L6088)):*
> ```Go
> 	statement_id   STRING NOT NULL
> );`,
> 	populate: func(ctx context.Context, p *planner, db catalog.DatabaseDescriptor, addRow func(...tree.Datum) error) (err error) {
> ```

nit: thoughts on implementing the `generate()` callback instead of `populate()`? 
___
*[`pkg/sql/sqlstats/outliers/outliers.go` line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80503#-N0XI19p3zlskhndcW0E:-N0XI19p3zlskhndcW0F:b-hi4fqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4249409b6f054459c67362c5d6790ac1bcd35421/pkg/sql/sqlstats/outliers/outliers.go#L40)):*
> ```Go
> // worrying about memory usage. See #79450.
> const (
> 	maxCacheSize = 10
> ```

How about use a cluster setting with default value of 10? 
___
*[`pkg/sql/sqlstats/outliers/outliers.go` line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80503#-N0XIvDo77BmWbF0H8JM:-N0XIvDo77BmWbF0H8JN:b7r0wmx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4249409b6f054459c67362c5d6790ac1bcd35421/pkg/sql/sqlstats/outliers/outliers.go#L89)):*
> ```Go
> 
> // ObserveTransaction notifies the registry of the end of a transaction.
> func (r *Registry) ObserveTransaction(sessionID uint128.Uint128, id uuid.UUID) {
> ```

Hmm why are we recording outliers only at the end of the transaction? and why do we need to buffer it in the `statements` field first? 

also minor nit: `s/id/txnID/`
___
*[`pkg/sql/sqlstats/outliers/outliers.proto` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80503#-N0XHqYgE5NCRD0j5MsV:-N0XHqYgE5NCRD0j5MsW:bxqvv2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/4249409b6f054459c67362c5d6790ac1bcd35421/pkg/sql/sqlstats/outliers/outliers.proto#L1)):*
> ```protobuf
> // Copyright 2021 The Cockroach Authors.
> ```

nit: 2022

**Review by matthewtodd — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80503)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng)

___
*[`pkg/sql/crdb_internal.go` line 6088 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80503#-N0XIZq12hML7SDfSEQU:-N0fWMFW8_1ndgs2Gdyx:b-lvljvg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4249409b6f054459c67362c5d6790ac1bcd35421/pkg/sql/crdb_internal.go#L6088)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: thoughts on implementing the `generate()` callback instead of `populate()`? 
</blockquote></details>

Thanks! Filed as #80639.
___
*[`pkg/sql/sqlstats/outliers/outliers.go` line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80503#-N0XI19p3zlskhndcW0E:-N0fXIfh8I1ysqkI3I_k:b-b2jb9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4249409b6f054459c67362c5d6790ac1bcd35421/pkg/sql/sqlstats/outliers/outliers.go#L40)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

How about use a cluster setting with default value of 10? 
</blockquote></details>

Yes, captured as #79450. The 10 here isn't what I imagine we'll end up with, just something to get us started.
___
*[`pkg/sql/sqlstats/outliers/outliers.go` line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80503#-N0XIvDo77BmWbF0H8JM:-N0fXhFB4f5BCSlXenf4:b-nr826g) ([raw file](https://github.com/cockroachdb/cockroach/blob/4249409b6f054459c67362c5d6790ac1bcd35421/pkg/sql/sqlstats/outliers/outliers.go#L89)):*
Good question! It follows from [this intention](https://cockroachlabs.atlassian.net/wiki/spaces/SQLOBS/pages/2322268382/Troubleshooting+slow+and+failed+executions+60#Detecting-slow-transaction-executions) in the feature brief:

> CockroachDB automatically records slow transaction executions similar to slow statement executions. When a transaction is ‘slow’ all its statements are captured (including individual statements that weren’t ‘slow’) accessible via SQL and the console.

The code here doesn't yet make any attempt at slow transaction detection, but I believe this general flow of information (buffer statements, analyze on/after transaction end) is what we'll need to do to support that use case.

(And good call on the `txnID` renaming, done.)
___
*[`pkg/sql/sqlstats/outliers/outliers.proto` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80503#-N0XHqYgE5NCRD0j5MsV:-N0fZDUsAVK86bzuWMga:bm9479d) ([raw file](https://github.com/cockroachdb/cockroach/blob/4249409b6f054459c67362c5d6790ac1bcd35421/pkg/sql/sqlstats/outliers/outliers.proto#L1)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: 2022
</blockquote></details>

Oops! Fixed.

File: pkg/sql/sqlstats/ssprovider.go [hunk 1]
```diff
@@ -183,6 +188,8 @@ type Provider interface {
 
 // RecordedStmtStats stores the statistics of a statement to be recorded.
 type RecordedStmtStats struct {
+	SessionID       uint128.Uint128
+	StatementID     uint128.Uint128
```
Comment (mgartner) [inline pkg/sql/sqlstats/ssprovider.go:192]: We should probably move `ClusterWideID` out of the `sql` package so you can use it here. I'm not sure the best place to move it to, maybe a new sub-package in `pkg/sql/cluster` and call it `cluster.ID`? I have a hunch that @ajwerner has a thoughtful opinion on the matter. If you do make the change, I recommend doing it in a separate commit before this one.

Comment (matthewtodd) [reply pkg/sql/sqlstats/ssprovider.go:192]: Talking live with @ajwerner, let's move it to `base`. We're thinking now about a better name, to distinguish it from `base.ClusterID`.

Comment (matthewtodd) [reply pkg/sql/sqlstats/ssprovider.go:192]: Or `pkg/sql/clusterwide`, or `pkg/sql/clusterunique`, calling it `clusterunique.ID`.

Comment (matthewtodd) [reply pkg/sql/sqlstats/ssprovider.go:192]: Underway in #80667.

**Review by Azhng — COMMENTED** [review]
> Reviewed 2 of 21 files at r1, 16 of 20 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80503)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, @matthewtodd, and @mgartner)

___
*[`pkg/sql/sqlstats/outliers/outliers_test.go` line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80503#-N0l8IZT4NeBp3vE3Kh1:-N0l8IZT4NeBp3vE3Kh2:bywbyyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1630c336147987050f970291b68c655c1a3cbcc/pkg/sql/sqlstats/outliers/outliers_test.go#L65)):*
> ```Go
> }
> 
> func TestOutliersDisabled(t *testing.T) {
> ```

very very small nit: i feel we can have a single `TestOutliers` test with some subtests (`t.Run()`) that tests regular/disabled/too fast test cases .
___
*[`pkg/sql/sqlstats/outliers/outliers_test.go` line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80503#-N0l93E--uEKPnpkyu33:-N0l93E0EBLfSixPNAGy:bbeemay) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1630c336147987050f970291b68c655c1a3cbcc/pkg/sql/sqlstats/outliers/outliers_test.go#L103)):*
> ```Go
> }
> 
> func TestOutliersConcurrentAccess(t *testing.T) {
> ```

Seems like the access to `registry` is completely serialized in this test?

**Review by matthewtodd — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80503)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, and @mgartner)

___
*[`pkg/sql/sqlstats/outliers/outliers_test.go` line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80503#-N0l8IZT4NeBp3vE3Kh1:-N0llkkOCGffIYg-1tB3:b-347h04) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1630c336147987050f970291b68c655c1a3cbcc/pkg/sql/sqlstats/outliers/outliers_test.go#L65)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

very very small nit: i feel we can have a single `TestOutliers` test with some subtests (`t.Run()`) that tests regular/disabled/too fast test cases .
</blockquote></details>

Yeah, I like that better, too. I didn't put too much effort into cleaning these tests up because I expected them to change a bunch, but I've gone ahead and pulled out the subtests as you suggested.
___
*[`pkg/sql/sqlstats/outliers/outliers_test.go` line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80503#-N0l93E--uEKPnpkyu33:-N0lljmRC6y6fZ1JDMoo:b-ljxc7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1630c336147987050f970291b68c655c1a3cbcc/pkg/sql/sqlstats/outliers/outliers_test.go#L103)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Seems like the access to `registry` is completely serialized in this test? 
</blockquote></details>

Sorry, "concurrent" wasn't the right word for me to have used here; it was misleading. This test is showing that the registry buffers statements per session id. I've renamed it.

**Review by Azhng — APPROVED** [review]
> Reviewed 28 of 59 files at r4, 1 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80503)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, and @mgartner)

**Pull request merged**